packages/server/__tests__/api/grants.test.js

```diff
@@ -378,4 +379,64 @@ describe('`/api/grants` endpoint', async () => {
             });
         });
     });
+    context('GET /api/grants/exportCSV', async () => {
+        it('produces correct column format', async () => {
+            // We constrain the result to a single grant that's listed in seeds/dev/ref/grants.js
+            const query = '?searchTerm=333816';
+            const response = await fetchApi(`/grants/exportCSV${query}`, agencies.own, fetchOptions.staff);
+
+            const expectedCsv = `Opportunity Number,Title,Viewed By,Interested Agencies,Status,Opportunity Category,Cost Sharing,Award Floor,Award Ceiling,Posted Date,Close Date,Agency Code,Grant Id,URL
+HHS-2021-IHS-TPI-0001,Community Health Aide Program:  Tribal Planning &amp; Implementation,,,posted,Discretionary,No,,500000,8/5/2021,9/6/2021,HHS-IHS,333816,https://www.grants.gov/web/grants/view-opportunity.html?oppId=333816
+`;
+
+            expect(response.statusText).to.equal('OK');
+            expect(response.headers.get('Content-Type')).to.include('text/csv');
+            expect(response.headers.get('Content-Disposition')).to.include('attachment');
+
+            expect(await response.text()).to.equal(expectedCsv);
+        });
+
+        it('limits output to 10k rows', async function () {
+            // First we insert 10k grants
+            const numToInsert = 10000;
+            const grantsToInsert = Array(numToInsert).fill(undefined).map((val, i, arr) => ({
+                status: 'inbox',
+                grant_id: String(-(i + 1)),
+                grant_number: String(-(i + 1)),
+                agency_code: 'fake',
+                cost_sharing: 'No',
+                title: `fake grant #${i + 1}/${arr.length} for test`,
+                cfda_list: 'fake',
+                open_date: '2022-04-22',
+                close_date: '2022-04-22',
+                notes: 'auto-inserted by test',
+                search_terms: '[in title/desc]+',
+                reviewer_name: 'none',
+                opportunity_category: 'Discretionary',
+                description: 'fake grant inserted by test',
+                eligibility_codes: '25',
+                opportunity_status: 'posted',
+                raw_body: 'raw body',
+            }));
+
+            // batchInsert is needed for large quantity of rows; it splits the insert into multiple
+            // queries
+            // TODO: is there a better way to test this that doesn't require actually inserting this
+            // many rows? Having a test this slow is not ideal. On my machine, take 2500ms
```

Comment:
only way i can think of is mocking knexjs db call or mocking the getGrants in db module

Reply:
Do we have any other API tests that mock this way you can point me at?

As far as I can tell, these API tests make requests against an actual running server in a completely separate process. The server behaves basically identically to normal except that based on `NODE_ENV=test` it connects to a different DB url (and it seems not to send login emails; can't really see where in the code prevents this from happening though since the testutils function hits the same login endpoints a user would).

Reply:
ah I see, then mocking is not an option. Sorry not catching that. Maybe we can lower the number of rows for this test. Seems that this is more of a stress testing than functional, unless we want to do both. In grants.gov i see around 6k of active grants

Reply:
The functional part of the test is that 1) the endpoint enforces a limit and doesn't try creating a massive export if there are more rows and 2) it correctly emits an error message at the bottom of the sheet if the exported data is incomplete.

But there is also a stress test component in that if someone changes the logic in `getGrants` in the future to do more complex queries, we would want to know that it doesn't break down at 10k the way it currently does at 100k (where I got errors in building a prepared statement with that many parameters).

If the current number is 6k, it seems believable to me that we could hit 10k at some point, though I think most users will probably have filtered that list down with eligibility codes and keywords.

Reply:
I decided 2500ms is not really acceptable in the test; decided to lower limit in test environment (via `NODE_ENV`) to 100 rows which makes the test run in under 100ms. Doesn't achieve the "stress test" goal above though.